refactor: actor search view model and its screen and add useful funct…

Noor-Serry opened the pull request:
add interaction listener make the search work in the view model 
also improve the UI  
# Pull Request Template

## Description
refactor search by actor view model and add interaction listener and enhance the base view model
---

## Changes Made
List the changes introduced in this pull request:

- actor search view model
- add interaction listener

---
## Screenshots (if applicable)
Include screenshots or GIFs to showcase the changes, especially if they impact the UI.
## 📺 Demo Video

<video src="https://github-production-user-asset-6210df.s3.amazonaws.com/95979643/465731541-39b5de91-521a-4651-acd5-b6846e992a1a.mp4" controls width="100%" />


---

## Checklist
Please ensure the following tasks are completed:

- [x] My code follows the code style of this project
- [x] Changes have been tested manually and verified.
- [x] PR includes at most one single feature.

---

## Additional Comments
Add any additional information or context about the pull request here.

Commit:
refactor: actor search view model and its screen and add useful function to base the base view model and add interaction listiner

ThoriaHamdy99 commented on viewModel/src/main/java/com/example/viewmodel/BaseViewModel.kt:
```diff
@@ -57,4 +56,10 @@ open class BaseViewModel<S, E>(initialState: S) : ViewModel() {
             }
         }
     }
+    protected fun launchDelayed(duration: Long, block: suspend CoroutineScope.() -> Unit): Job {
```
Why we will use this function

ThoriaHamdy99 reviewed (changes requested):
The change of base view model shouldn't be merged now and in this branch as you are doing the actor screen.

Commit:
Merge branch 'fix/remote-data-source-error' into refactor/search-by-actor-view-model

Commit:
Merge branch 'develop' of github.com:Amsterdam-Team/Aflami into refactor/search-by-actor-view-model

Commit:
refactor: implement debounced search in SearchByActorScreen

This commit refactors the search functionality in `SearchByActorScreen`
to use a debounced Flow for handling user input.

Commit:
Merge branch 'develop' of github.com:Amsterdam-Team/Aflami into refactor/search-by-actor-view-model

# Conflicts:
#	viewModel/src/main/java/com/example/viewmodel/BaseViewModel.kt
#	viewModel/src/main/java/com/example/viewmodel/searchByActor/SearchByActorViewModel.kt

ThoriaHamdy99 reviewed (commented):
Resolve conflicts

Commit:
refactor: make tryToExecute return Job

The `tryToExecute` function in `BaseViewModel` now returns a `Job`. This allows for cancellation of the coroutine if needed.

The `SearchByActorViewModel` now accepts a `DispatcherProvider` in its constructor, consistent with other ViewModels.

Additionally, an import path in `AflamiDatabase.kt` and a package declaration in `InstantConverter.kt` were corrected.

ThoriaHamdy99 commented on viewModel/src/main/java/com/example/viewmodel/searchByActor/SearchByActorViewModel.kt:
```diff
@@ -6,58 +6,75 @@ import com.example.entity.Movie
-            state
-                .map { it.query }
+            queryFlow
                 .debounce(300)
```
Why you didn't use the search string in the whole screen state. Why you are using new mutable state --> 
    private val queryFlow = MutableStateFlow("") to debounce on ?
    I need to understand your opinion here.

Noor-Serry replied:
I check how you implement the seache in the search by country view model but you did not implement any delay between each search character that is not good you making a lot of call while the user typing .
the best solution I found it with this way using private flow and emit some value I changed the preivous way that I found it in Asia repo  
it was work but when the user delete the character very fast its gone to do unexpected behavior
